File: sdk/src/main/java/com/trackingplan/client/sdk/TrackingplanConfig.java
// Copyright (c) 2021 Trackingplan
package com.trackingplan.client.sdk;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

import com.trackingplan.shared.TrackingplanConfigBuilder;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Immutable class.
 * Wraps the shared TrackingplanConfig and adds Android-specific configuration.
 */
final public class TrackingplanConfig {

    public static final int MAX_REQUEST_BODY_SIZE_IN_BYTES =
            com.trackingplan.shared.TrackingplanConfig.MAX_REQUEST_BODY_SIZE_IN_BYTES;

    public final static TrackingplanConfig EMPTY = new TrackingplanConfig();

    // Shared configuration (contains common fields)
    private final com.trackingplan.shared.TrackingplanConfig sharedConfig;

    // Android-specific fields
    private final Map<String, String> customContext;
    private final boolean backgroundObserver;

    @VisibleForTesting
    public static Builder newConfig(@NonNull String tpId) {
        return new Builder(tpId);
    }

    private TrackingplanConfig() {
        // Use empty() factory for sentinel/placeholder values
        this.sharedConfig = com.trackingplan.shared.TrackingplanConfig.Companion.empty();
        this.customContext = Collections.emptyMap();
        this.backgroundObserver = true;
    }

    private TrackingplanConfig(
            com.trackingplan.shared.TrackingplanConfig sharedConfig,
            Map<String, String> customContext,
            boolean backgroundObserver
    ) {
        this.sharedConfig = sharedConfig;
        this.customContext = Collections.unmodifiableMap(new HashMap<>(customContext));
        this.backgroundObserver = backgroundObserver;
    }

    @NonNull
    public String getTpId() {
        return sharedConfig.getTpId();
    }

    public boolean isBackgroundObserverEnabled() {
        return backgroundObserver;
    }

    public boolean isDebugEnabled() {
        return sharedConfig.getDebug();
    }

    boolean isTestingEnabled() {
        return sharedConfig.getTesting();
    }

    public boolean isDryRunEnabled() {
        return sharedConfig.getDryRun();
    }

    public String getSourceAlias() {
        return sharedConfig.getSourceAlias();
    }

    @NonNull
    public String getEnvironment() {
        return sharedConfig.getEnvironment();
    }

    @NonNull
    public Map<String, String> customContext() {
        return customContext;
    }

    @NonNull
    public Map<String, String> customDomains() {
        return sharedConfig.getProviderDomains();
    }

    @NonNull
    public Map<String, String> tags() {
        return sharedConfig.getTags();
    }

    @NonNull
    public String getTracksEndPoint() {
        return sharedConfig.getTracksEndpoint();
    }

    @NonNull
    public String getConfigEndPoint() {
        return sharedConfig.getConfigEndpoint();
    }

    /**
     * Creates a new TrackingplanConfig instance with updated tags.
     *
     * @param newTags The tags to add or set. Must not be null.
     * @param replace If true, replaces all existing tags with the new tags.
     *                If false, merges new tags with existing tags (new values overwrite existing ones for same keys).
     * @return A new TrackingplanConfig instance with updated tags
     */
    @NonNull
    public TrackingplanConfig withTags(@NonNull Map<String, String> newTags, boolean replace) {
        com.trackingplan.shared.TrackingplanConfig updatedSharedConfig =
                sharedConfig.withTags(newTags, replace);
        return new TrackingplanConfig(
                updatedSharedConfig,
                this.customContext,
                this.backgroundObserver
        );
    }

    @Override
    @NonNull
    public String toString() {
        return "TrackingplanConfig{" +
                "tpId='" + getTpId() + '\'' +
                ", environment='" + getEnvironment() + '\'' +
                ", sourceAlias='" + getSourceAlias() + '\'' +
                ", dryRun='" + isDryRunEnabled() + '\'' +
                ", debug='" + isDebugEnabled() + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TrackingplanConfig that = (TrackingplanConfig) o;
        return backgroundObserver == that.backgroundObserver
                && customContext.equals(that.customContext)
                && sharedConfig.equals(that.sharedConfig);
    }

    @Override
    public int hashCode() {
        return Objects.hash(sharedConfig, customContext, backgroundObserver);
    }

    public static class Builder {

        private final TrackingplanConfigBuilder sharedBuilder;
        private final Map<String, String> customContext = new HashMap<>();
        private boolean backgroundObserver = true;

        public Builder(String tpId) {
            if (tpId.isEmpty()) {
                throw new IllegalArgumentException("Parameter tpId cannot be empty");
            }
            sharedBuilder = new TrackingplanConfigBuilder().tpId(tpId);
        }

        public Builder configEndPoint(@NonNull String configEndPoint) {
            sharedBuilder.configEndpoint(configEndPoint);
            return this;
        }

        @VisibleForTesting
        public Builder customContext(@NonNull Map<String, String> context) {
            this.customContext.clear();
            this.customContext.putAll(context);
            return this;
        }

        public Builder customDomains(@NonNull Map<String, String> customDomains) {
            sharedBuilder.providerDomains(customDomains);
            return this;
        }

        public Builder enableDebug() {
            sharedBuilder.debug(true);
            return this;
        }

        public Builder enableTesting() {
            sharedBuilder.testing(true);
            return this;
        }

        public Builder enableDryRun() {
            sharedBuilder.dryRun(true);
            return this;
        }

        public Builder environment(@NonNull String environment) {
            sharedBuilder.environment(environment);
            return this;
        }

        @VisibleForTesting
        public Builder disableBackgroundObserver() {
            this.backgroundObserver = false;
            return this;
        }

        @Deprecated
        public Builder ignoreContext() {
            // Ignored
            return this;
        }

        public Builder sourceAlias(@NonNull String alias) {
            sharedBuilder.sourceAlias(alias);
            return this;
        }

        public Builder tags(@NonNull Map<String, String> tags) {
            sharedBuilder.tags(tags);
            return this;
        }

        public Builder tracksEndPoint(@NonNull String tracksEndPoint) {
            sharedBuilder.tracksEndpoint(tracksEndPoint);
            return this;
        }

        @NonNull
        public TrackingplanConfig build() {
            com.trackingplan.shared.TrackingplanConfig sharedConfig = sharedBuilder.build();
            return new TrackingplanConfig(
                    sharedConfig,
                    customContext,
                    backgroundObserver
            );
        }
    }
}

File: shared/src/commonMain/kotlin/com/trackingplan/shared/TrackingplanConfig.kt
// Copyright (c) 2021 Trackingplan
package com.trackingplan.shared

/**
 * Immutable shared configuration for Trackingplan SDK.
 * This class contains all common configuration fields shared between platforms.
 *
 * Use [TrackingplanConfigBuilder] to create instances of this class.
 */
@ConsistentCopyVisibility
data class TrackingplanConfig internal constructor(
    val tpId: String,
    val environment: String = "PRODUCTION",
    val sourceAlias: String = defaultSourceAlias(),
    val tags: Map<String, String> = emptyMap(),
    val providerDomains: Map<String, String> = emptyMap(),
    val debug: Boolean = false,
    val testing: Boolean = false,
    val dryRun: Boolean = false,
    val tracksEndpoint: String = DEFAULT_TRACKS_ENDPOINT,
    val configEndpoint: String = DEFAULT_CONFIG_ENDPOINT
) {

    /**
     * Creates a new immutable config with updated tags.
     *
     * @param newTags The tags to add or set.
     * @param replace If true, replaces all existing tags with the new tags.
     *                If false, merges new tags with existing tags (new values overwrite existing ones for same keys).
     * @return A new TrackingplanConfig instance with updated tags
     */
    fun withTags(newTags: Map<String, String>, replace: Boolean = false): TrackingplanConfig {
        val mergedTags = if (replace) {
            newTags
        } else {
            tags + newTags
        }
        return copy(tags = mergedTags)
    }

    /**
     * Generates the sampling rate configuration URL.
     */
    fun sampleRateUrl(): String {
        return "${configEndpoint}config-${tpId}.json"
    }

    companion object {
        const val MAX_REQUEST_BODY_SIZE_IN_BYTES = 100 * 1024
        const val DEFAULT_TRACKS_ENDPOINT = "https://eu-tracks.trackingplan.com/v1/"
        const val DEFAULT_CONFIG_ENDPOINT = "https://config.trackingplan.com/"

        /**
         * Creates an empty/sentinel config. Used for placeholder values.
         */
        // TODO: Find another way to represent empty/sentinel config without bypassing tpId validation
        fun empty(): TrackingplanConfig = TrackingplanConfig(tpId = "")
    }
}

/**
 * Builder for TrackingplanConfig with fluent API.
 * The build() method returns an immutable TrackingplanConfig instance.
 *
 * @throws IllegalArgumentException if tpId is empty when build() is called
 */
class TrackingplanConfigBuilder {

    private var tpId: String = ""
    private var environment: String = "PRODUCTION"
    private var sourceAlias: String = defaultSourceAlias()
    private var tags: Map<String, String> = emptyMap()
    private var providerDomains: Map<String, String> = emptyMap()
    private var debug: Boolean = false
    private var testing: Boolean = false
    private var dryRun: Boolean = false
    private var tracksEndpoint: String = TrackingplanConfig.DEFAULT_TRACKS_ENDPOINT
    private var configEndpoint: String = TrackingplanConfig.DEFAULT_CONFIG_ENDPOINT

    fun tpId(tpId: String) = apply { this.tpId = tpId }
    fun environment(environment: String) = apply { this.environment = environment }
    fun sourceAlias(alias: String) = apply { this.sourceAlias = alias }
    fun tags(tags: Map<String, String>) = apply { this.tags = tags.toMap() }
    fun providerDomains(domains: Map<String, String>) = apply { this.providerDomains = domains.toMap() }
    fun debug(enabled: Boolean) = apply { this.debug = enabled }
    fun testing(enabled: Boolean) = apply { this.testing = enabled }
    fun dryRun(enabled: Boolean) = apply { this.dryRun = enabled }
    fun tracksEndpoint(endpoint: String) = apply {
        this.tracksEndpoint = normalizeEndpoint(endpoint)
    }
    fun configEndpoint(endpoint: String) = apply {
        this.configEndpoint = normalizeEndpoint(endpoint)
    }

    @Throws(IllegalArgumentException::class)
    fun build(): TrackingplanConfig {
        require(tpId.isNotEmpty()) { "Parameter tpId cannot be empty" }
        require(!dryRun || debug) { "DryRun mode requires Debug mode to be enabled" }
        return TrackingplanConfig(
            tpId = tpId,
            environment = environment,
            sourceAlias = sourceAlias,
            tags = tags,
            providerDomains = providerDomains,
            debug = debug,
            testing = testing,
            dryRun = dryRun,
            tracksEndpoint = tracksEndpoint,
            configEndpoint = configEndpoint
        )
    }

    private fun normalizeEndpoint(endpoint: String): String {
        return if (endpoint.endsWith("/")) endpoint else "$endpoint/"
    }
}

/**
 * Platform-specific default source alias.
 */
expect fun defaultSourceAlias(): String
